src/KeyRegistry.sol

```diff
@@ -0,0 +1,299 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.19;
+
+import {Ownable2Step} from "openzeppelin/contracts/access/Ownable2Step.sol";
+import {IdRegistry} from "./IdRegistry.sol";
+
+contract KeyRegistry is Ownable2Step {
+    /**
+     *  @notice State enumeration for a key in the registry. During migration, an admin can change
+     *          the state of any fids key from NULL to ADDED or ADDED to NULL. After migration, an
+     *          fid can change the state of a key from NULL to ADDED or ADDED to REMOVED only.
+     *
+     *          - NULL: The key is not in the registry.
+     *          - ADDED: The key has been added to the registry.
+     *          - REMOVED: The key was added to the registry, but is now removed.
+     */
+    enum KeyState {
+        NULL,
+        ADDED,
+        REMOVED
+    }
+
+    /**
+     *  @notice Data about a key.
+     *
+     *  @param state   The current state of the key.
+     *  @param scheme  The manner in which the key should be used.
+     */
+    struct KeyData {
+        KeyState state;
+        uint32 scheme;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                                 ERRORS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @dev Revert if a key violates KeyState transition rules.
+    error InvalidState();
+
+    /// @dev Revert if owner attempts a bulk add or reset after the migration grace period.
+    error Unauthorized();
+
+    /// @dev Revert if owner calls migrateKeys more than once.
+    error AlreadyMigrated();
+
+    /// @dev Revert if migration batch input arrays are not the same length.
+    error InvalidBatchInput();
+
+    /*//////////////////////////////////////////////////////////////
+                                 EVENTS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @dev Emit an event when an admin or fid adds a new key.
+     *
+     * @param fid       The fid associated with the key.
+     * @param scheme    The type of the key.
+     * @param key       The key being registered. (indexed as hash)
+     * @param keyBytes  The bytes of the key being registered.
+     * @param metadata  Metadata about the key.
+     */
+    event Add(uint256 indexed fid, uint32 indexed scheme, bytes indexed key, bytes keyBytes, bytes metadata);
+
+    /**
+     * @dev Emit an event when an fid removes an added key.
+     *
+     * @param fid       The fid associated with the key.
+     * @param key       The key being registered. (indexed as hash)
+     * @param keyBytes  The bytes of the key being registered.
+     */
+    event Remove(uint256 indexed fid, bytes indexed key, bytes keyBytes);
```

Comment:
Why does Remove have `bytes indexed key` and `bytes keyBytes` if both are set to `key` in its only usage: https://github.com/farcasterxyz/contracts/pull/257/files#diff-ffa5866d21942cfa6a42043498fbb08d5ef20a0d5d97ba107f6d94fb33713e59R207 ?